apps/prairielearn/src/middlewares/AuthzAccessMismatch.tsx

```diff
@@ -0,0 +1,225 @@
+                  {permission.value ? 'Yes' : 'No'}
+                </span>
+              ) : (
+                permission.value
+              )}
+            </td>
+            <td>
+              {permission.type === 'boolean' ? (
+                <span class={`badge ${permission.authnValue ? 'bg-success' : 'bg-danger'}`}>
+                  {permission.authnValue ? 'Yes' : 'No'}
+                </span>
+              ) : (
+                permission.authnValue
+              )}
+            </td>
+          </tr>
+        ))}
+      </tbody>
+    </table>
+  );
+};
+
+export function AuthzAccessMismatch({
+  errorMessage,
+  oneOfPermissionKeys,
+  authzData,
+  authnUser,
+  authzUser,
+}: {
+  errorMessage: string;
+  oneOfPermissionKeys: (keyof PageContext['authz_data'])[];
+  authzData: PageContext['authz_data'];
+  authnUser: StaffUser;
+  authzUser: StaffUser | null;
+}) {
+  const clearEffectiveUserCookies = () => {
+    removeCookieClient(['pl_requested_uid', 'pl2_requested_uid']);
+    removeCookieClient(['pl_requested_course_role', 'pl2_requested_course_role']);
+    removeCookieClient(['pl_requested_course_instance_role', 'pl2_requested_course_instance_role']);
+    removeCookieClient(['pl_requested_date', 'pl2_requested_date']);
+    setCookieClient(['pl_requested_data_changed', 'pl2_requested_data_changed'], 'true');
+    window.location.reload();
+  };
+
+  const permissions: PermissionData[] = PermissionsMeta.map((permission) => {
+    return {
+      authnValue:
+        (authzData as any)['authn_' + permission.key] ??
+        (permission.type === 'string' ? '' : false),
+      value: authzData[permission.key] ?? (permission.type === 'string' ? '' : false),
+      ...permission,
+    };
+  });
+
+  const [oneOfPermissions, allOtherPermissions] = _.partition(permissions, (permission) =>
+    oneOfPermissionKeys.includes(permission.key),
+  );
+
+  // Only show the permissions that are different between authn and authz
+  const otherPermissions = allOtherPermissions.filter(
+    (permission) => permission.authnValue !== permission.value,
+  );
+
+  return (
+    <main id="content" class="container">
+      <div class="card mb-4">
+        <div class="card-header bg-danger text-white">Insufficient access</div>
+        <div class="card-body">
+          <h2 class="mb-3 h4">Effective user has insufficient access</h2>
+          <p>
+            The
+            {authzUser ? (
+              <OverlayTrigger
+                overlay={
+                  <Tooltip>
+                    {authzUser.name} ({authzUser.uid})
+                  </Tooltip>
+                }
+              >
+                <button type="button" class="btn btn-link link-secondary p-0 mx-1 align-baseline">
+                  current effective user
+                </button>
+              </OverlayTrigger>
+            ) : (
+              ' current user '
+            )}
```

Comment:
I can't shake the feeling that using all these tooltips is simply worse than just presenting things inline.

> The current effective user (foo@example.com) does not have access to this page, but your account (bar@example.com) does. This page requires either course preview access or student data view access.

Reply:
I reworked things in #12597